core/src/main/java/org/apache/calcite/plan/RelOptUtil.java

```diff
@@ -2199,13 +2199,49 @@ public static boolean equal(
       RelDataType type2,
       Litmus litmus) {
     if (!areRowTypesEqual(type1, type2, false)) {
-      return litmus.fail("Type mismatch:\n{}:\n{}\n{}:\n{}",
+      return litmus.fail("Type mismatch:\n{}:\n{}\n{}:\n{}\ndifference:\n{}",
           desc1, type1.getFullTypeString(),
-          desc2, type2.getFullTypeString());
+          desc2, type2.getFullTypeString(),
+          getFullTypeDifferenceString(type1, type2));
     }
     return litmus.succeed();
   }
 
+  public static String getFullTypeDifferenceString(RelDataType source,
+      RelDataType target) {
+    if (source == target) {
+      return "";
+    }
+
+    if (source.getFieldCount() > target.getFieldCount()) {
+      return "Type mismatch: the target type has less fields than the source type";
```

Comment:
I'm afraid `Type mismatch: the target type has less fields than the source type` provides very little information to the developer on what is wrong.

Reply:
The error message is a hint to tell what may be wrong. From my experience, it is enough.

Reply:
Agreed with @vlsi , it may be helpful to provide the field count. 
If so, there is no need to distinguish the cases for source count > target count and source count < target count.